### ISSUE TITLE
Jetson TX2 Target

### ISSUE THREAD

**Issue by scotthong** [jcuda/jcuda-main#20]
Hi,
I've successfully updated jcuda to support Jetson TX2. I hope this can be adopted in the future release.
Here are a list of changes I've made to make it work on Jetson TX2.
1. jcuda-common/JCudaParentPOM/pom.xml: Add a new "linux-aarch64" profile
```
        <profile>
            <id>linux-aarch64</id>
            <activation>
                <os>
                    <family>unix</family>
                    <arch>aarch64</arch>
                </os>
            </activation>
            <properties>
                <jcuda.os>linux</jcuda.os>
                <jcuda.arch>aarch64</jcuda.arch>
                <jcuda.ext>so</jcuda.ext>
                <jcuda.prefix>lib</jcuda.prefix>
            </properties>
        </profile>      
```

2.  jcuda-common/JCudaCommon_CMake.txt@line 26: Add "/targets/aarch64-linux/lib" to the PATH_SUFFIXES
Change from:
```
    PATH_SUFFIXES /lib/x64 /targets/x86_64-linux/lib)
```
To:
```
    PATH_SUFFIXES /lib/x64 /targets/x86_64-linux/lib /targets/aarch64-linux/lib)
```
3. jcuda-common/JCudaCommon_CMake.txt@line 62 add the following JCUDA_ARCH
```
    elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
      if(CMAKE_SIZEOF_VOID_P EQUAL 8)
        set(JCUDA_ARCH "aarch64")
      endif()
```
4. LibUtils.java@84 add "AARCH64" to the ArchType enum
```
    public static enum ArchType
    {
        PPC, PPC_64, SPARC, X86, X86_64, ARM, ARM64, AARCH64, MIPS, MIPS64, RISC, UNKNOWN
    }
````
5. LibUtils.java@547 add "aarch64" to the "calculateArch" function
```
    if (osArch.startsWith("aarch64"))
    {
        return ArchType.AARCH64;
    }
```

**Comment by jcuda** [comment]
Thanks for this contribution!

Someone had created the binaries for TK1 a while ago, as also mentioned in [this forum thread](https://forum.byte-welt.net/t/jcuda-on-the-nvidia-jetson-tk1-dev-board/8949), but the build process was still (even more) manual back then. Having this properly integrated in the Maven/CMake toolchain would really be great!

I'm a bit confused, though: Which version did you base this on? Just because you mentioned the `jcuda-common/JCudaParentPOM/pom.xml` file. During the Mavenization of JCuda, this file has essentially been moved to the `jcuda-parent` project, at https://github.com/jcuda/jcuda-parent/blob/master/pom.xml 

However, from what I can see, the changes *should* not interfere with existing targets and profiles, because they all seem to clearly be triggered by the `aarch64` part.

I'll integrate them ASAP.

Would you be able to contribute the corresponding binaries (i.e. the `.lib` files) after the update? (I'm not familiar with the Jetson world, but assume that they'd rather be offered as a dedicated download from the website, and *not* as a Maven artifact, right?)

**Comment by jcuda** [comment]
The changes have been added in...

- `jcuda-parent`, updating the parent POM: https://github.com/jcuda/jcuda-parent/commit/491d8ce8358e5ded68e7c8e632d24b61774444ca
- `jcuda-common`, updating the CMake file: https://github.com/jcuda/jcuda-common/commit/c146d9770dc9774199f46c5be82cdfde3a39734c
`jcuda`, updating the `LibUtils` class: https://github.com/jcuda/jcuda/commit/d909b6f9dcbc2952da5d409b4e16ebe61421bea6

It would be great if you could verify that it works. 

(Although the main and most pressing task right now is to get the 0.9.0 update out, and this should be unaffected by these changes...)

**Comment by scotthong** [comment]
These changes are based on 0.8.0. 
I will upload precompiled binaries after verifying the changes works on my Jetson TX2. I am assuming that the branch I should be verifying is 0.9.0. Please let me know where I can upload the precompiled binaries. Thanks for the great work to enable Java developers to take advantage of the cuda libraries.

**Comment by scotthong** [comment]
I just checkout the latest master to build jcuda from source. cmake-gui can create the build file  successfully without error. However, the build failed with the error messages attached below. It seems that the latest jcuda-0.9.0 depends on cuda-9. According to the official response from nVidia, cuda-9 for Jetson TX2 is not available yet and the official release date is not given. It will be probably early 2018. I also build the 0.9.0-RC00 branch on my Ubuntu 14.04 with cuda-8.0 with the same error messages.

https://devtalk.nvidia.com/default/topic/1024821/when-jetpack-with-cuda-9-for-jetson-tx2-/

As far as the pre-build binaries of JCuda 0.9.0 for Jetson TX2 goes, it will probably going to take awhile before I have a chance to upgrade my Jetson TX2 to the latest release in the middle of the development cycle. JCuda 0.8.0 works for the development I am working on. Most likely, I won't be able to provide the pre-build binaries in a timely manner. At least, for the people working on Jetson TX2, they should have no problem building JCuda 0.9.0 from source if the "aarch64" toolchain is part of the 0.9.0 release.

```
[ 21%] Building CXX object jcuda/JCudaDriverJNI/bin/CMakeFiles/JCudaDriver.dir/src/JCudaDriver.cpp.o
In file included from /home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:29:0:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver_common.hpp:154:2: error: ‘CUDA_LAUNCH_PARAMS’ does not name a type
  CUDA_LAUNCH_PARAMS cudaLaunchParams;
  ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp: In function ‘bool initCUDA_LAUNCH_PARAMSData(JNIEnv*, jobject, CUDA_LAUNCH_PARAMSData*)’:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:1143:2: error: ‘CUDA_LAUNCH_PARAMS’ was not declared in this scope
  CUDA_LAUNCH_PARAMS &cudaLaunchParams = data->cudaLaunchParams;
  ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:1143:22: error: ‘cudaLaunchParams’ was not declared in this scope
  CUDA_LAUNCH_PARAMS &cudaLaunchParams = data->cudaLaunchParams;
                      ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:1143:47: error: ‘CUDA_LAUNCH_PARAMSData {aka struct CUDA_LAUNCH_PARAMSData}’ has no member named ‘cudaLaunchParams’
  CUDA_LAUNCH_PARAMS &cudaLaunchParams = data->cudaLaunchParams;
                                               ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp: In function ‘bool getOptionValue(JNIEnv*, jobject, CUjit_option, void*&)’:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:1208:28: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
             value = (void*)v;
                            ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:1220:28: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
             value = (void*)iv;
                            ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp: In function ‘jint Java_jcuda_driver_JCudaDriver_cuLaunchCooperativeKernelNative(JNIEnv*, jclass, jobject, jint, jint, jint, jint, jint, jint, jint, jobject, jobject)’:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:4557:21: error: ‘cuLaunchCooperativeKernel’ was not declared in this scope
   nativeKernelParams);
                     ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp: In function ‘jint Java_jcuda_driver_JCudaDriver_cuLaunchCooperativeKernelMultiDeviceNative(JNIEnv*, jclass, jobjectArray, jint, jint)’:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:4580:2: error: ‘CUDA_LAUNCH_PARAMS’ was not declared in this scope
  CUDA_LAUNCH_PARAMS *launchParamsListNative = new CUDA_LAUNCH_PARAMS[len];
  ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:4580:22: error: ‘launchParamsListNative’ was not declared in this scope
  CUDA_LAUNCH_PARAMS *launchParamsListNative = new CUDA_LAUNCH_PARAMS[len];
                      ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:4580:51: error: ‘CUDA_LAUNCH_PARAMS’ does not name a type
  CUDA_LAUNCH_PARAMS *launchParamsListNative = new CUDA_LAUNCH_PARAMS[len];
                                                   ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:4589:52: error: ‘CUDA_LAUNCH_PARAMSData {aka struct CUDA_LAUNCH_PARAMSData}’ has no member named ‘cudaLaunchParams’
   launchParamsListNative[i] = launchParamsDatas[i].cudaLaunchParams;
                                                    ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:4591:121: error: ‘cuLaunchCooperativeKernelMultiDevice’ was not declared in this scope
 unchCooperativeKernelMultiDevice(launchParamsListNative, (unsigned int)numDevices, (unsigned int)flags);
                                                                                                       ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp: In function ‘jint Java_jcuda_driver_JCudaDriver_cuFuncSetAttributeNative(JNIEnv*, jclass, jobject, jint, jint)’:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:4653:72: error: ‘cuFuncSetAttribute’ was not declared in this scope
  int result = cuFuncSetAttribute(nativeHFunc, nativeAttrib, nativeValue);
                                                                        ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp: In function ‘jint Java_jcuda_driver_JCudaDriver_cuStreamWaitValue64Native(JNIEnv*, jclass, jobject, jobject, jlong, jint)’:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:6873:85: error: ‘cuStreamWaitValue64’ was not declared in this scope
  int result = cuStreamWaitValue64(nativeStream, nativeAddr, nativeValue, nativeFlags);
                                                                                     ^
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp: In function ‘jint Java_jcuda_driver_JCudaDriver_cuStreamWriteValue64Native(JNIEnv*, jclass, jobject, jobject, jlong, jint)’:
/home/nvidia/TX2SSD/Projects/jcuda/jcuda/JCudaDriverJNI/src/JCudaDriver.cpp:6899:86: error: ‘cuStreamWriteValue64’ was not declared in this scope
  int result = cuStreamWriteValue64(nativeStream, nativeAddr, nativeValue, nativeFlags);
                                                                                      ^
jcuda/JCudaDriverJNI/bin/CMakeFiles/JCudaDriver.dir/build.make:62: recipe for target 'jcuda/JCudaDriverJNI/bin/CMakeFiles/JCudaDriver.dir/src/JCudaDriver.cpp.o' failed
make[2]: *** [jcuda/JCudaDriverJNI/bin/CMakeFiles/JCudaDriver.dir/src/JCudaDriver.cpp.o] Error 1
CMakeFiles/Makefile2:87: recipe for target 'jcuda/JCudaDriverJNI/bin/CMakeFiles/JCudaDriver.dir/all' failed
make[1]: *** [jcuda/JCudaDriverJNI/bin/CMakeFiles/JCudaDriver.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by jcuda** [comment]
Yes, the error messages simply indicate that the new structures and functions of CUDA 9 have not been found. So this should be resolved when CUDA 9 is available for Jetson. (I wasn't aware that there is a delay between the main release and the Jetson release). 

So until then, I'll leave this issue open and have an eye on the thread that you linked to.

Thanks again.

**Comment by scotthong** [comment]
JetPack 3.2 - Developer Preview has just released. This is a non-production Developer Preview.

https://developer.nvidia.com/embedded/jetpack-notes

**Comment by jcuda** [comment]
Great. I'll hardly be able to test this myself soon, but if you encounter any issues, please let me know.

**Comment by jcuda** [comment]
There don't seem to be so many people targeting Jetson/Jetpack. If someone wants to contribute or has any issue with building the binaries, this could be discussed in a new issue.
